### ISSUE TITLE
virtual service rewrite doing urldecoding

### ISSUE THREAD

**Issue by ceastman-ibm** [envoyproxy/envoy#13147]
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: virtual service rewrite doing urldecoding

*Description*:
>What issue is being seen? Describe what should be happening instead of
the bug, for example: Envoy should not crash, the expected value isn't
returned, etc.

Setup a virtual service rewrite to an external authority and then pass an url that is encoded, see virtual service yaml at the end of issue

https://hostname.com/api/apim/crn%3Av1%3Astaging%3Apublic%3A

is getting sent to endpoint.com as:

/apim/crn:v1:staging:public:

we need it to remain url encoded

/apim/crn%3Av1%3Astaging%3Apublic%3A

*Repro steps*:
> Include sample requests, environment, etc. All data and inputs
required to reproduce the bug.
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: proxy-istio
spec:
  hosts:
    - hostname.com
  gateways:
  - proxy-istio
  http:
    - match:
        - uri:
            prefix: /api
      rewrite:
        authority: endpoint.com
        uri: " "
      route:
        - destination:
            host: endpoint.com
            port:
              number: 443

I logged this issue with the istio team but they recommended logging it to envoyproxy instead: https://github.com/istio/istio/issues/27301

>**Note**: The [Envoy_collect tool](https://github.com/envoyproxy/envoy/blob/master/tools/envoy_collect/README.md)
gathers a tarball with debug logs, config and the following admin
endpoints: /stats, /clusters and /server_info. Please note if there are
privacy concerns, sanitize the data prior to sharing the tarball/pasting.

*Admin and Stats Output*:
>Include the admin output for the following endpoints: /stats,
/clusters, /routes, /server_info. For more information, refer to the
[admin endpoint documentation.](https://www.envoyproxy.io/docs/envoy/latest/operations/admin)

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Config*:
>Include the config used to configure Envoy.

*Logs*:
>Include the access logs and the Envoy logs.

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Call Stack*:
> If the Envoy binary is crashing, a call stack is **required**.
Please refer to the [Bazel Stack trace documentation](https://github.com/envoyproxy/envoy/tree/master/bazel#stack-trace-symbol-resolution).

**Comment by yanavlasov** [comment]
My suspicion is that path canonicalization decodes %-encoded characters.

**Comment by ceastman-ibm** [comment]
@yanavlasov and it only seems to happen when the rewrite is in the virtual service. if we dont have the rewrite section the url that is encoded is just passed onto the destination.

**Comment by ceastman-ibm** [comment]
@yanavlasov Is there a timeline for when this will get looked at?

**Comment by yanavlasov** [comment]
Not yet. We just cut a release out and now we can figure out if there resources to do this work.

**Comment by udaymadu** [comment]
@yanavlasov Do we have any update on it and would like to know which release would it be part of.

**Comment by udaymadu** [comment]
I am just checking to see any one working on it.

**Comment by udaymadu** [comment]
@yanavlasov is there a timeline when will it get fixed.

**Comment by udaymadu** [comment]
Is anyone actively looking at it.

**Comment by yanavlasov** [comment]
We have the first part in review right now, PR #13808 This PR will not solve this issue, however. I think it will need some additional work. I do not have ETA for you right now.

**Comment by udaymadu** [comment]
Thanks for the update.

**Comment by udaymadu** [comment]
@yanavlasov Have you defined any milestone for it or do we defined any ETA as it became severe now.

**Comment by yanavlasov** [comment]
No milestones, sorry. There is nobody working on the issue that raised here specifically, just some related parts. If you need this fixed soon you need to find someone who can do the development.

**Comment by PiotrSikora** [comment]
This is effectively a duplicate of #6589, since Envoy is incorrectly forwarding the sanitized URI (used to perform internal access control checks and routing) instead of the original URI.

**Comment by udaymadu** [comment]
Is there any action around it in upcoming releases about it.

**Comment by lildeadprince** [comment]
Is there any updates for the issue?

For my case the problem was that my application were using path parameters routing  (e.g. `/user/:someStringIdentifier/actionHistory`) and in some rare cases values of path parameters were containing the slash characters ("/", ASCII 2F hex).

Also the rewrite was configured in Istio for the service, so Envoy were decoding those slashes and not encoding them back.
Based on the results, as far as I understand, it performs like `encodeURI(rewrite(decodeURI(url)))` instead of something like

```js
joinPath(
  rewrite(
    splitURI(url).map(decodeURIComponent)
  )
  .map(encodeURIComponent)
)
```

### tldr:
I expect Envoy to follow the "Example 2" from the [W3C URI Recommendations](https://www.w3.org/Addressing/URL/4_URI_Recommentations.html):

> The URIs
> http://www.w3.org/albert/bertram/marie-claude
> 
> and
> http://www.w3.org/albert/bertram%2Fmarie-claude
> 
> are NOT identical, as in the second case the encoded slash does not have hierarchical significance.
